File: 78. Subsets/Subsets.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        if not nums:
            return []
        output = []
        self.helper(nums, output, 0, [])
        return output
    
    def helper(self, nums: List[int], output: List[List[int]], index: int, cur: List[int]) -> None:
        # Add a copy of the current subset to output
        output.append(cur[:])
        
        # Explore each element starting from index
        for i in range(index, len(nums)):
            cur.append(nums[i])           # Include current element
            self.helper(nums, output, i + 1, cur)  # Recurse with next index
            cur.pop()                    # Backtrack by removing last element

File: 78. Subsets/Subsets.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> output = new ArrayList<>();
        backtrack(nums, output, 0, new ArrayList<>());
        return output;
    }

    private void backtrack(int[] nums, List<List<Integer>> output, int start, List<Integer> path) {
        output.add(new ArrayList<>(path)); // Add a copy of the current subset
        for (int i = start; i < nums.length; i++) {
            path.add(nums[i]); // Include nums[i]
            backtrack(nums, output, i + 1, path); // Recurse
            path.remove(path.size() - 1); // Backtrack (exclude nums[i])
        }
    }
}